File: Daily/PY/Leetcode211-添加与搜索单词.py
class Node:
    def __init__(self):
        self.child = [None] * 26
        self.isend = False 

class WordDictionary:

    def __init__(self):
        self.root = Node()

    def addWord(self, word: str) -> None:
        p = self.root
        for i in range(len(word)):
            c = ord(word[i]) - ord('a')
            if not p.child[c]:
                p.child[c] = Node()
            p = p.child[c]
        p.isend = True 

    def search(self, word: str) -> bool:
        p = self.root
        def dfs(p,word,idx):
            if idx == len(word): return p.isend
            if word[idx] == '.':
                for i in range(26):
                    if p.child[i] and dfs(p.child[i],word,idx+1):
                        return True
            else:
                c = ord(word[idx]) - ord('a')
                if p.child[c]:
                    return dfs(p.child[c],word,idx+1)
            return False
        return dfs(p,word,0)

File: Leetcode/sbgo/Leetcode211-添加与搜索单词.go
package sbgo

type Node struct {
	Child [26]*Node
	Isend bool
}

type WordDictionary struct {
	Root *Node
}

func Constructor() WordDictionary {
	return WordDictionary{&Node{}}
}

func (p *WordDictionary) AddWord(word string) {
	t := p.Root
	for _, c := range word {
		c -= 'a'
		if t.Child[c] == nil {
			t.Child[c] = &Node{}
		}
		t = t.Child[c]
	}
	t.Isend = true
}

func (p *WordDictionary) Search(word string) bool {

	var dfs func(*Node, int) bool
	dfs = func(p *Node, idx int) bool {
		if idx == len(word) {
			return p.Isend
		}
		if word[idx] == '.' {
			for i := 0; i < 26; i++ {
				if p.Child[i] != nil && dfs(p.Child[i], idx+1) {
					return true
				}
			}
		} else {
			c := word[idx] - 'a'
			if p.Child[c] != nil {
				return dfs(p.Child[c], idx+1)
			}
		}
		return false
	}
	return dfs(p.Root, 0)
}
